blocks/main_loop.py

```diff
@@ -178,28 +199,52 @@ def _run_iteration(self):
         self.algorithm.process_batch(batch)
         self.status.iterations_done += 1
         self._run_extensions('after_batch', batch)
-        self._check_finish_training()
+        self._check_finish_training('batch')
         return True
 
     def _run_extensions(self, method_name, *args):
         for extension in self.extensions:
             extension.dispatch(method_name, *args)
 
-    def _check_finish_training(self):
+    def _check_finish_training(self, level):
+        """Checks whether the current training should be terminated.
+
+        Parameters
+        ----------
+        level : {'epoch', 'batch'}
+            The level at which this check was performed. In some cases, we
+            only want to quit after completing the remained of the epoch.
+
+        """
         # In case when keyboard interrupt is handled right at the end of
         # the iteration the corresponding log record can be found only in
         # the previous row.
         if (self.log.current_row.training_finish_requested or
-                self.log.current_row.keyboard_interrupt_received or
-                self.log.previous_row.keyboard_interrupt_received):
+                self.log.current_row.batch_interrupt_received or
+                self.log.previous_row.batch_interrupt_received):
+            raise TrainingFinish
+        if (level == 'epoch' and
+                (self.log.current_row.epoch_interrupt_received or
```

Comment:
Don't know how I could miss that, but of course that does not work. Assume you make a record `epoch_interrupt_received` at the iteration 234, and your epoch ends at iteration 400. Neither the current nor the previous row can contain this record.

I guess we on the keyboard interrupts we need to make records into both log and status. Into the log, in order to be able to reconstruct what happened later, into the status, in order to for the information to be easily accessible.

Same change could be applied to `batch_interrupt_received` to avoid looking into the previous row.